src/Vault.sol

```diff
@@ -0,0 +1,175 @@
+        require(collateralRequired < deposits[msg.sender], "Insufficient collateral!");
+
+        debts[msg.sender] += debtAmount;
+        bool sent = debt.transfer(msg.sender, debtAmount);
+        require(sent, "Borrow failed!");       
+        emit Borrow(address(debt), msg.sender, debtAmount);
+    }
+
+    ///@notice Users repay their debt, in debt asset terms
+    ///@dev This covers partial and full repayment
+    ///@param debtAmount Amount of debt asset to repay
+    function repay(uint debtAmount) external {
+        debts[msg.sender] -= debtAmount;
+
+        bool sent = debt.transferFrom(msg.sender, address(this), debtAmount);
+        require(sent, "Repayment failed!");       
+        emit Repay(address(debt), msg.sender, debtAmount);   
+    }
+
+
+    ///@notice Users withdraw their deposited collateral
+    ///@dev This covers partial and full withdrawal; checks for spare capacity before initiating withdrawal
+    ///@param collateralAmount Amount of collateral asset to withdraw
+    function withdraw(uint collateralAmount) external {       
+        uint userDebt = debts[msg.sender];                          // caching this saves an SLOAD
+        uint userDeposit = deposits[msg.sender];                   // caching this saves an SLOAD
+
+        if (userDebt > 0){
+            uint collateralRequired = getCollateralRequired(userDebt);
+            uint spareDeposit = userDeposit - collateralRequired;
+            require(collateralAmount < spareDeposit, "Collateral unavailable!");
+        }
+        
+        deposits[msg.sender] -= collateralAmount;
+        bool sent = collateral.transfer(msg.sender, collateralAmount);
+        require(sent, "Withdraw failed!");
+        emit Withdraw(address(collateral), msg.sender, collateralAmount);            
+    }
+
+    ///@notice Price is returned as an integer extending over its decimal places
+    ///@dev Calculates collateral required to support existing debt position at current market prices 
+    ///@param debtAmount Amount of debt asset to support
+    /*Note: Example calculation using collateralLevel (e.g. DAI/WETH):
+    *       baseCollateralRequired = 1 WETH (1e18), collateralLevel = 80% (8e17), scalarFactor = 10**18  (DAI/WETH is 18dp)
+    *       1 WETH (1e18) * 80% (8e17) / 10**18 = 8e(18+17) / 10**18 = 8e17  => 80% of 1 WETH
+    *       Since we want to have a buffer, we will DIVIDE by collateralLevel: 1/0.8 = 1.25 (20% buffer, 80% collateralization level)
+    */
+    function getCollateralRequired(uint debtAmount) public view returns(uint) {
+        (,int price,,,) = priceFeed.latestRoundData();
+        uint baseCollateralRequired = debtAmount * uint(price) / scalarFactor;
+        uint bufferCollateralRequired = (baseCollateralRequired * scalarFactor) / collateralLevel; 
+        return bufferCollateralRequired = scaleDecimals(bufferCollateralRequired, collateral.decimals(), debt.decimals());
```

Comment:
Each one of these `.decimals()` is a CALL to an external contract, and expensive at 2100+ gas. You should read the decimals in the constructor and store them, since they won't change.